### PR TITLE
FX pass to move input mutations into submodule

### PR DESCRIPTION
**Pull request by bdhirsh**
When functionalization is turned on in AOT Autograd, we want to hide input mutations in the graph so that the backend compiler doesn't need to worry about seeing `copy_()` ops in the graph. This PR does that by hiding it in an opaque submodule.

Right now this logic happens after the partitioning, and we're relying on partitioning to always leave the `copy_()` nodes in the forward graph (which... probably needs some more testing, but I think is fine?).

I added light testing for this pass by including it in the existing `test_functionalization.py` tests, but I'm planning to try hooking this into the torchbench suite, which will let us get rid of this code: https://github.com/pytorch/torchdynamo/blob/5040d49795dde35f0112e27a6744015d44318deb/torchdynamo/optimizations/training.py#L59

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85844
* #85875
* #85681
* #82601
* __->__ #82602
* #82771
* #83265

### CODE REVIEW

**Comment by ezyang** [comment]
It would be really nice to see a before/after print

File: functorch/functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:246]: So if you do this after, `partition_fn` must be able to deal with mutations in `fx_g`. Can it, @Chillee ?

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:171]: submodule graph, you mean?

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:174]: nit: this only remaps one input at a time

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:176]: how come?

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:217]: I know we debated a bit possible representations, but this particular rep is actually not one that I had been thinking about. Mutation be in a submodule *in the original* graph is not great, because it still means the outer graph is *not functional*! (If you call a mutating function inside a graph, that makes your graph mutating.) I feel like there's an obligation for this submodule call to (somehow) not live in the Graph itself, as that makes it vulnerable to DCE again.

Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:217]: Hmmmm I see. The alternative is probably just to leave it out of the `fx.Graph` (but still keep it in the `GraphModule`), and then manually cal the submodule later on in AOTAutograd, after the compiled forward gets executed? That doesn't seem too bad.

I guess my thought was that anything that operates on the `fx.Graph` (DCE + compilers) would just see a custom submodule and know to treat it as an opaque object. But that's probably not totally right - we can't really enforce that a graph pass won't move the submodule around earlier in the graph (which would be wrong).

Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:217]: although, if I want to be able to invoke the submodule separately outside of the graph, I'll need to somehow get the inputs for the submodule - probably by updating the original graph to make them additional outputs. Extra complexity (since I'll also need to remove the pytree stuff to do that), but doesn't seem too bad.

File: functorch/functorch/_src/aot_autograd.py
Comment (bdhirsh) [reply functorch/functorch/_src/aot_autograd.py:246]: I re-remembered that it can't haha, so I'll have to change this. the min-cut partition function code calls `eliminate_dead_code()`, so we want the `copy_()` ops to be hidden away before then.

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [reply functorch/functorch/_src/partitioners.py:217]: You can always write passes so they operate correctly in the presence of mutation. But we don't want to force it on pass writers. Opaque submodules at final lowering are unlikely to be doing heavy optimization, so it's easier to deal with arbitrary side effects. But even, e.g., finding fusion groups, is currently not correct with mutating things (though mostly this is because of DCE calls). But the problem here is you're adding in this module all the way at the beginning,  before *all* of the optimization passes, and so you're forcing them to, at the very least, know not to remove your module.

Extra outputs is what I would expect to see if it's external.

Comment (ezyang) [reply functorch/functorch/_src/partitioners.py:217]: You could also have it not actually be external, but stored on the GraphModule and not actually part of the graph. But passes would need to know to propagate it.

Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:217]: hmm yeah, having it actually be external feels better. I'll try that.

File: functorch/functorch/_src/partitioners.py
Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:176]: yep this was me misunderstanding FX nodes again, removing

File: test/test_functionalization.py
Comment (bdhirsh) [inline test/test_functionalization.py:36]: The changes in this file are just minor QoL changes

**Comment by bdhirsh** [comment]
Ok, I beefed up this PR (so mutations get taken out before parititioning) and added better testing. @Chillee lmk what you think of the AOTAutograd changes.

I also dumped my tests in `test_compile_cache.py` since it was the only file I found that directly tests `aot_function()`, but I can move it somewhere else if preferred.

I think what we probably want is that once this lands, we should turn this + re-inplacing on by default in the benchmark testing on torchbench + timm + hugging face (cc @anijain2305). I'm still working on getting that set up so I can actually run it locally and confirm that nothing breaks.

**Comment by bdhirsh** [comment]
It looks like this pass isn't playing well with the partitioning code - it fails `python test/test_pythonkey.py TestAOTAutograd.test_batchnorm` when functionalization is toggled on. I think the partitioning code needs some beefing up but I'm taking a closer look.

File: functorch/functorch/_src/partitioners.py
Comment (bdhirsh) [inline functorch/functorch/_src/partitioners.py:98]: @Chillee let me know how.. suspicious you think this looks. I can't think of a cleaner way to do this; it seems like the partitioning code uses info from the pytree spec on the original joint graph to carve out the forward/backward, and I don't think we want to update that info when we add extra outputs in the forward graph 🤔

**Comment by ezyang** [comment]
got any... before after prints? :)

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [reply functorch/functorch/_src/partitioners.py:98]: So there is a bigger question here, which is that when AOTAutograd is wrapped around a chunk of code that makes mutations (or worse, metadata changes) to some inputs, what exactly are we supposed to output in this case.

It seems to me like there are two cases. The easy case is when the mutated inputs don't support grad. Then you can just do the mutation outside of the custom autograd function and you end up with a variable that doesn't support gradients, but that's what the original program would have done anyway, no problem.

The harder case is when the mutated input *does* support grad (because it's like an inplace or something). Custom autograd function can't really deal with this. I'm not sure if there's an easy way to get it to work. also cc @albanD

**Comment by bdhirsh** [comment]
I added an expect test  in `test_compile_cache.py` (maybe I should move it somewhere else) - it also doesn't include a before, which I'll add:

```
        # Function takes in 4 total tensors, mutates 1 and 3, returns 2 and 4
        # takes in / returns lists to check py-tree handling
        def fn(tensor1, lst, tensor2):
            a = tensor1
            b = lst[0]
            c = lst[1]
            d = tensor2

            a_view = a.view(-1)
            a_view.add_(1)
            c_view = c.view(-1)
            c_view.add_(1)

            return [b, b], d

         out = functorch.make_fx(functorch.experimental.functionalize(fn))(                              
            torch.ones(2), [torch.ones(2), torch.ones(2)], torch.ones(2))

# before graph:
print(out.graph)

graph():
    %tensor1_1 : [#users=2] = placeholder[target=tensor1_1]
    %lst_1 : [#users=1] = placeholder[target=lst_1]
    %lst_2 : [#users=2] = placeholder[target=lst_2]
    %tensor2_1 : [#users=1] = placeholder[target=tensor2_1]
    %view_default : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%tensor1_1, [-1]), kwargs = {})
    %add_tensor : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%view_default, 1), kwargs = {})
    %view_default_1 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%lst_2, [-1]), kwargs = {})
    %add_tensor_1 : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%view_default_1, 1), kwargs = {})
    %view_default_2 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%add_tensor, [2]), kwargs = {})
    %view_default_3 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%add_tensor_1, [2]), kwargs = {})
    %copy__default : [#users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%tensor1_1, %view_default_2), kwargs = {})
    %copy__default_1 : [#users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%lst_2, %view_default_3), kwargs = {})
    return [lst_1, lst_1, tensor2_1]

# move mutations into submodule
move_input_mutations_into_submodule(out)

# after graph:
print(out.graph)

graph():
    %tensor1_1 : [#users=1] = placeholder[target=tensor1_1]
    %lst_1 : [#users=1] = placeholder[target=lst_1]
    %lst_2 : [#users=1] = placeholder[target=lst_2]
    %tensor2_1 : [#users=1] = placeholder[target=tensor2_1]
    %view_default : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%tensor1_1, [-1]), kwargs = {})
    %add_tensor : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%view_default, 1), kwargs = {})
    %view_default_1 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%lst_2, [-1]), kwargs = {})
    %add_tensor_1 : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%view_default_1, 1), kwargs = {})
    %view_default_2 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%add_tensor, [2]), kwargs = {})
    %view_default_3 : [#users=1] = call_function[target=torch.ops.aten.view.default](args = (%add_tensor_1, [2]), kwargs = {})
    return [view_default_2, view_default_3, lst_1, lst_1, tensor2_1]


# print the new mutation submodule
print(out.input_mutation_submodule.graph)

graph():
    %tensor1_1 : [#users=1] = placeholder[target=tensor1_1]
    %view_default_2 : [#users=1] = placeholder[target=view_default_2]
    %copy__default : [#users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%tensor1_1, %view_default_2), kwargs = {})
    %lst_2 : [#users=1] = placeholder[target=lst_2]
    %view_default_3 : [#users=1] = placeholder[target=view_default_3]
    %copy__default_1 : [#users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%lst_2, %view_default_3), kwargs = {})



```

File: functorch/functorch/_src/partitioners.py
Comment (albanD) [reply functorch/functorch/_src/partitioners.py:98]: In custom Function, we allow the input to be modified IIF the input is not a view or the custom Function has a single output.
Similarly, if the custom Function returns a view, we mark it such that it cannot be modified inplace in a differentiable manner.

So we do support some cases and raise errors for all the others.

Comment (ezyang) [reply functorch/functorch/_src/partitioners.py:98]: Right. But this level of support would be insufficient for arbitrarily whacky partial graphs from dynamo, where any of those situations could occur.

**Comment by ezyang** [comment]
How do I know how many outputs are "fake" and how to get to the real outputs?

File: functorch/functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:243]: when does x evaluate to false? Seems fishy...

Comment (ezyang) [reply functorch/functorch/_src/aot_autograd.py:243]: oh is mutated_inputs_map not actually a dict. Don't call it a map lol

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:250]: Given the code in AOTAutograd, it doesn't seem like you make use of the fact that the mutation module is a submodule at all, since you don't have a `call_submodule` call anymore. In that case, might as well just not add it as a submodule and just return it from this function. Partitioning in a true sense.

File: functorch/functorch/_src/aot_autograd.py
Comment (Chillee) [inline functorch/functorch/_src/aot_autograd.py:186]: Doesn't need to be done in this PR, but there's enough state here that we should probably be refactoring it out into a separate object.

File: functorch/functorch/_src/aot_autograd.py
Comment (Chillee) [inline functorch/functorch/_src/aot_autograd.py:266]: It seems like we're assuming that only the forwards pass will have its inputs mutated? Is this true?

File: functorch/functorch/_src/aot_autograd.py
Comment (Chillee) [inline functorch/functorch/_src/aot_autograd.py:284]: I think we're going to need to pass this input mutation submodule to backends as well, potentially.

File: functorch/functorch/_src/partitioners.py
Comment (Chillee) [reply functorch/functorch/_src/partitioners.py:98]: > it seems like the partitioning code uses info from the pytree spec on the original joint graph to carve out the forward/backward

I think we should refactor that logic out a bit to be more explicit :)

**Review by Chillee — COMMENTED** [review]
Doing a first pass over the code. Sorry, I haven't been following this PR along (although I should be...), so might be retreading on some previous convos.

File: functorch/functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:289]: It's lucky that AOTAutograd is the only place using this, because the calling convention is pretty complicated and I definitely would not be wanting to force arbitrarily many clients to have to deal with this by hand. I guess this was the appeal of putting it in codegen: if you did this in codegen, the return signature of the module doesn't actually change and all of the extra bookkeeping is hidden from the user (at the cost of needing to know that you have to preserve the codegen attribute across passes, and at the cost of needing to know that if you want to retrace without seeing the `copy_` calls, you have to actually run the interpreter. This second one is not such a big deal, since @SherlockNoMad is about to require you to do that anyway--see my NACK on https://github.com/pytorch/pytorch/pull/82755 )

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:220]: @SherlockNoMad is the API ready for checking preconditions yet? The precondition here is that the graph is functionalized and it's all torch.ops calls

File: functorch/functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/functorch/_src/partitioners.py:238]: Do you wanna assert that this is actually an input node?

**Review by ezyang — APPROVED** [review]
I'd really like @Chillee to give some suggestions on any cleanup we should do for AOTAutograd

**Comment by Chillee** [comment]
> I'd really like @Chillee to give some suggestions on any cleanup we should do for AOTAutograd

Offhand

1. Factor out the "compile cache" state within the autograd.Function into a separate object.
2. Figure out a better way of passing semantic information about how many inputs/outputs to the partitioner.
3. Add some more types/rename some things to better variable names.
4. Maybe preserve more semantic naming information on inputs.

File: functorch/functorch/_src/partitioners.py
Comment (albanD) [reply functorch/functorch/_src/partitioners.py:98]: That is correct. But I don't think it is technically possible to support these! (that's another problem of Macro-custom Function!)

File: functorch/functorch/_src/partitioners.py
Comment (SherlockNoMad) [reply functorch/functorch/_src/partitioners.py:220]: Something like this? 

https://github.com/pytorch/pytorch/pull/83134

File: functorch/functorch/_src/aot_autograd.py
Comment (bdhirsh) [reply functorch/functorch/_src/aot_autograd.py:266]: Yep (I commented that assumption somewhere else in this file). I was talking to @soulitzer a few weeks ago about it, and it sounds like it shouldn't be valid for the backward to mutate its program inputs (?).

File: functorch/functorch/_src/aot_autograd.py
Comment (bdhirsh) [reply functorch/functorch/_src/aot_autograd.py:284]: hmm. What do you think is the benefit of sending it to backends - do you think there's potential for a backend to optimize those `copy_()` in some ways?

My mental model was basically:
(1) Create a giant purely functional graph, let the backend go wild and optimize it
(2) If there are any input mutations afterwards, just run the eager code to patch up the inputs (I assumed this isn't worth optimizing, but maybe that's wrong)

File: functorch/functorch/_src/partitioners.py
Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:98]: > In custom Function, we allow the input to be modified IIF the input is not a view or the custom Function has a single output.

> Right. But this level of support would be insufficient for arbitrarily whacky partial graphs from dynamo, where any of those situations could occur.

So, my original goal when writing this PR was that I saw that `torch.dynamo` currently skips `AOTAutograd` entirely [when it detects input mutations](https://github.com/pytorch/torchdynamo/blob/3f2decb5b91d43931606efc72b34f581be839d9d/torchdynamo/optimizations/training.py#L68), and we want to reduce the number of subgraphs that dynamo is forced to fall back on.

Based on the above discussion, it feels wrong to land this PR exactly as is: if `torch.autograd.Function` fundamentally can't support input mutations some cases (that Alban enumerated above), then it seems like AOTAutograd should also properly error in those cases, and dynamo should correctly know to fall back in those cases.

What do you all think of the following?
(1) Update this PR so that AOTAutograd raises an error in the cases Alban pointed out above (if the input we're mutating is a view, and the forward's return type is mult-output)
(2) Some point later, update dynamo to use the same criteria when deciding whether to run the fallback and skip AOTAUtograd

Comment (ezyang) [reply functorch/functorch/_src/partitioners.py:98]: cc @anijain2305 @jansel but I think erroring in the backend is too late, assuming `raise_on_backend_error` is on False (which it's not by default), it will still nuke optimizations for the entire graph.

It seems to me like we should AOTAutograd the functional component only, and then strap on a bunch of (unfused) mutations at the tail end that don't get wrapped in AOTAutograd.

File: functorch/functorch/_src/aot_autograd.py
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:266]: makes sense

File: functorch/functorch/_src/partitioners.py
Comment (bdhirsh) [reply functorch/functorch/_src/partitioners.py:97]: Coming back to this - using AOTAutograd only on the functional components, and moving the input mutations elsewhere, fully outside of AOTAutograd, makes sense to me.

Models that mutate metadata actually seem like they might not be that uncommon: For example, quantized models like `resnet50_quantized_qat` have running_min/running_max state that get mutated and resized (the ATen op that does this mutating is `_fused_moving_avg_obs_fq_helper`). And the same is probably true for batch norm.

I think my biggest concern is around the API, and how to make it in a way that is (potentially) not BC-breaking for AOTAutograd users.

Some design questions:
(1) Should it be on the callee (dynamo here) to manually call the extra mutation logic? Or should AOTAutograd still arrange for that to happen automatically (I think we probably prefer for this all to still happen transparently, given the complicated calling convention, but let me know if someone thinks differently).
(2) How acceptable is it to change the return type of `aot_function()`?



Brainstorming - the easiest way would probably be to have AOTAutograd always return an `nn.module`, that transparently calls the compiled function (which is an `autograd.Function` object that can't handle certain input mutations, and then manually performs the metadata mutations. E.g. something like this:
```
    class AOTWrapper(nn.Module):
            ...
        def forward(self, *args, **kwargs):
            # compiled_f is the original autograd.Function object that can't handle certain input mutations
            out = compiled_f(*args, **kwargs)
            self.mutate_inputs_submodule(original_inputs, updated_inputs)
```

**Comment by bdhirsh** [comment]
I'm pushing some more on this based on https://github.com/pytorch/pytorch/issues/85036, since adding an epilogue to AOTAutograd should unblock a few models that were previously hitting dynamo's fallback. This isn't ready for review yet though - waiting to sanity check some passes tests first.

**Comment by bdhirsh** [comment]
@Chillee @ezyang this should be ready for another round. Some high level questions/notes:

(a) What do you all think of the partitioner calling convention changes?

(b) Should functionalization (and the same epilogue infra) be refactored to also run in the aot_dispatch_base() case? It looks like.. functionalization doesn’t even run there today (I guess I'm surprised that it isn't breaking anything?)

(c) There are a bunch of test failures that are due to fake tensors not being turned on (we need to know requires_grad-ness of the outputs to know what to mark properly in the `autograd.function`, but that information is wrong if you're using `make_fx()` with normal tracing, since we store a `TensorMeta` object in the FX graph that never gets its requires_grad field properly fixed up later)

**Comment by ezyang** [comment]
I'm deferring to @Chillee for this.

**Comment by bdhirsh** [comment]
@Chillee (and also @wconstab, since you're familiar with the partitioner) friendly bump.

I still need to rebase this though, on top of the recent partitioner changes.

File: functorch/_src/aot_autograd.py
Comment (jjsjann123) [inline functorch/_src/aot_autograd.py:510]: So after the pass, all mutation would be in opaque mutation epilogue and backends lose visibility there.
We are missing out on fusion opportunities here.

Would we be able to opt-in to inline the mutation epilogue back into the main graph?

Comment (bdhirsh) [reply functorch/_src/aot_autograd.py:510]: > So after the pass, all mutation would be in opaque mutation epilogue

TBC, this is only for captured graphs with input mutations. Intermediate mutations in a graph would have already been removed by functionalization.

The idea in this PR as its stands is that instead of the backend seeing an operator like:
```
native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor(a!) running_mean, Tensor(a!) running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor)
```

It'll see a purely functional version, that returns the updated inputs instead of mutating them directly:
```
native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor running_mean, Tensor running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor, Tensor, Tensor)
```

And if mutated inputs to that operator happen to correspond to graph inputs (which is true for running_mean/running_var variables for batch norm), then at the end of the graph we'll have an epilogue that copies the "updated inputs" back to the original inputs:
```
def compiled_fn(inpt1, inpt2):
     # ... first run the entirely functional compiled function
     outs, mutated_inpt1, mutated_inpt2 = real_compiled_fn(inpt1, inpt2)
     inpt1.copy_(mutated_inpt1)
     inpt2.copy_(mutated_inpt2)
     return outs
```

> Would we be able to opt-in to inline the mutation epilogue back into the main graph?

I remember @Chillee bringing this up before - we probably can? Although for now, this PR doesn't do that and just ensures that we do the "correct" thing first.

It's also not clear to me - why would this prevent fusions? The main disadvantage as I see it is that you can end up using more memory - you have to keep the buffer for both the original input and the updated input around while the compiled function is running.

**Review by wconstab — APPROVED** [review]
looks correct to me. thanks @bdhirsh

**Comment by bdhirsh** [comment]
I'm actually going to close this PR and create a fresh one. Thanks for the stamp and sorry about that Will. This is mostly because:

(1) Ed's "trace with functionalization in one pass" PR has landed and changed how we want this PR to work pretty substantially - we no longer have to back `copy_()` and `as_strided_()` into the graph, and then analyze the graph. We can just directly detect the input mutation case from inside of AOTAutograd and add the proper handling

(2) There are a bunch of other edge cases that are worth thinking about more holistically. Ed has a great doc on them here: https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit?usp=sharing

File: functorch/_src/aot_autograd.py
Comment (jjsjann123) [reply functorch/_src/aot_autograd.py:510]: Thanks for the explanation, I think we are mostly on the same page here. I'm mostly asking for mutation on inputs, since that's what normalization layers uses for running stats update.

> It's also not clear to me - why would this prevent fusions?

It's the epilogue in-place copies that we are missing out. Since those are cheap and easy to handle in normalization kernels. i.e. If we keep those in the epilogue, they won't be visible to fuser backend.

**Pull request closed without merging**
